### ISSUE TITLE
win_viewport is under documented and possibly glitchy

### ISSUE THREAD

**Issue by fredizzimo** [neovim/neovim#23609]
### Problem

Neovide is implementing smooth scrolling using the `win_viewport` event, but there are a few problems. These should have been caught before merging https://github.com/neovim/neovim/pull/19270, and as a part of the Neovide organization I want to start by apologizing for us pressuring you to merge that in a hurry, before we even had an implementation on the Neovide side.

Now, we have two different implementation the one in the released version 0.10.4 and the one in my https://github.com/neovide/neovide/pull/1790. The pull request first took an alternative approach of using the `grid_scroll` event, and assuming that it represented a real scroll, when the area was the complete visible portion of the screen. I never personally saw any problems with that approach, but @Kethku reported some problems. None one else reported any problems either, so my plan was to communicate with you and see if we could use that approach, perhaps by making it more official and include some extra flags to the event that tells if it's a scroll or not. However, soon after https://github.com/neovim/neovim/pull/19270, was merged, and Neovide master started using that, so I decided to follow and update my branch to the `scroll_delta` as well. But it quickly turned out that both the Neovide release and my branch has the same problems. For more context, see https://github.com/neovide/neovide/pull/1790, especially towards the end, where I also explain these issues in a slightly different way.

The immediate problem is that the updates are not atomic when scrolling, this is an example of two batches of UI events (for a slightly more detailed and different explanation see https://github.com/neovide/neovide/pull/1790)
```
grid_line {
    grid: 4,
    row: 0,
    column_start: 126,
    cells: [...] }
win_viewport { 
    grid: 2,
    top_line: 27.0,
    bottom_line: 113.0,
    current_line: 65.0,
    current_column: 0.0,
    line_count: Some(456.0), 
    scroll_delta: Some(3.0) }
flush
```
```
grid_scroll { 
    grid: 2,
    top: 0,
    bottom: 66,
    left: 0,
    right: 137,
    rows: 3,
    columns: 0 }
grid_line {
    grid: 2,
    row: 63,
    column_start: 0,
    cells: [...] })
grid_line { 
    grid: 2,
    row: 64, 
    column_start: 0,
    cells: [...] }
grid_line {
    grid: 2,
    row: 65,
    column_start: 0,
    cells: [...] }
win_viewport {
    grid: 2,
    top_line: 27.0,
    bottom_line: 113.0,
    current_line: 65.0,
    current_column: 0.0,
    line_count: Some(456.0),
    scroll_delta: Some(0.0) }
cursor_goto {
    grid: 2,
    row: 38,
    column: 4 }
flush
```
As you can see the first batch sends a flush with a new viewport. But the problem here is that the viewport has not yet been updated, and Neovide has no way of knowing that, so it just displays what was on the screen at that location.  The viewport is then redrawn in the second event. So, the whole sequence looks like the screen is flickering due to displaying the wrong contents temporarily.

The real problem here is that the `win_viewport` event is underspecified. Nowhere in the documentation does it say how it's supposed to work. One interpretation, and the one we are using is to assume that `win_viewport` acts as an internal flush of the previous viewport, and all the `grid_line` events following it will apply to the new viewport. But this interpretation is clearly wrong, since `win_viewport` is one of the last events a batch, confirmed both in practice and by reading the Neovim source code in `ui_flush`.  

From the code it furthermore looks like `win_viewport` is only sent once per flush batch, which could mean that `win_viewport` is supposed to apply to the whole batch. This works, but it's slightly inconvenient since it means that we can't update our internal grid representation in the same order as the events arrive, instead we have to buffer the events and wait for the viewport event before doing anything.

There are other interpretations as well that allows multiple `win_viewport` per batch.

None of these interpretations matches the behaviour in the example above, unless we assume that `win_viewport` has nothing do with the scrolling, it just moves the pointer to the next destination of the buffer to be updated. In that case everything is consistent. But it's still unclear if `win_viewport` applies to the current batch, or if it applies only to the following batches.

But if `win_viewport` also is a notification that the buffer has been scrolled, which I think it's supposed to be since the documentation clearly says that the `grid_scroll` is not a real scroll indication, then it really needs to part of the same batch that updates the destination to avoid glitches. 

My current workaround in https://github.com/neovide/neovide/pull/1790, is to assume that `win_viewport` applies to the next batch, and simply ignore any scrolling until the next batch, which I'm expecting to always be sent right after. This seems to work in most cases, but I don't believe it works in all of them. Especially if the current position is updated and scrolling happens at the same time, for example when deleting something partially inside and partially outside the screen.

I have read the source code a bit, to try to determine what's actually going on. The most important parts seem to be in `ui_flush` and `update_screen`.  But especially this comment
```c
// Commands that scroll a window change w_topline and must call
// check_cursor() to move the cursor into the visible part of the window, and
// call redraw_later(wp, UPD_VALID) to have the window displayed by update_screen()
// later.
```
So, it looks like scrolling, will cause a delayed redrawing of the screen, which I what I observed in practice as well, and seems to support my current workaround.  But it only holds if scrolling is the only type of update, in other cases the screen might be redrawn immediately, which breaks my workaround.

I don't really have any suggestions on how to solve this, I don't know the inner workings of Neovim enough, and things gets really complex when we start considering updates that are a hybrid of scrolling and modifications, like deleting/inserting lines, or expanding/collapsing folds. But to me it looks like the best would be to encode all that in the `grid_scroll` event and add some more fields describing what type of scroll it is and how many lines actually scrolled, even outside of the screen.  Then Neovide could act on just that, like my original implementation, and forget that `win_viewport` exists.  That event could still be useful for other applications though.

### Steps to reproduce

Use Neovim through the latest release of Neovide 0.10.4. 

1. Create a mapping that scrolls down and up a few lines for example `nmap <Down> 3<C-e>`  and  `nmap <Up> 3<C-y>`
2. Open a buffer with some text
3. Scroll down a bit so that you can scroll up later
4. Fold some lines, so that the screen is partially folded
5. Scroll up/down with the mapping
6. Observe that the screen suddenly jumps

### Expected behavior

The scrolling should work without glitches.

### Neovim version (nvim -v)

NVIM v0.9.0

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Arch Linux

### Terminal name/version

Neovide 0.10.4

### $TERM environment variable

N/A

### Installation

pacman

**Comment by vhakulinen** [comment]
I might be encountering similar issue with gnvim. This issue is fairly long, and its quite late atm here so I'll just add my (hopefully complementing) explanation:

When nvim sends `win_viewport` event, it reports the `scroll_delta` value a few batches into "future". I.e. the `scroll_delta` value is for a screen state that is only transferred after a few more batches (a few `flush` events) later.

**Comment by fredizzimo** [comment]
I created a pull request that adresses the issue #24182

**Comment by justinmk** [comment]
> I don't really have any suggestions on how to solve this ... But to me it looks like the best would be to encode all that in the `grid_scroll` event and add some more fields describing what type of scroll it is and how many lines actually scrolled, even outside of the screen. Then Neovide could act on just that, like my original implementation, and forget that `win_viewport` exists. That event could still be useful for other applications though.

Looks like that solution wasn't needed in https://github.com/neovim/neovim/pull/24182 ?

**Comment by fredizzimo** [comment]
No, it wasn't but #24182, is not perfect, it's mainly a band aid to get things working.

Incorporating the viewport to grid_scroll would be much more flexible, since we could optionally smooth scroll things like inserting and pasting lines, folding/unfolding and so on, in addition to normal scrolling. But that also means that `grid_scroll` need to be much more specified so that the user can distinguish between different types of "scrolling"

The main reason for my solution not being perfect is that you receive the the new grid lines first and then `win_viewport`, so some kind of buffer is needed on the user side. Doing it any other way would be tricky, since due to folds and wrapped lines and such all the information is not available before the full update.
